### PR TITLE
Instruction Tuning Improvements

### PR DESCRIPTION
**Pull request by le1nux**
# What does this PR do?

This PR ..

## General Changes
* ..

## Breaking Changes
* .. 

## Checklist before submitting final PR
- [ ] My PR is minimal and addresses one issue in isolation
- [ ] I have merged the latest version of the target branch into this feature branch
- [ ] I have reviewed my own code w.r.t. correct implementation, missing type hints, proper documentation, etc.
- [ ] I have run a sample config for model training
- [ ] I have checked that all tests run through (`python tests/tests.py`)
- [ ] I have updated the internal changelog (`CHANGELOG_DEV.md`)

### CODE REVIEW

**Review by le1nux — COMMENTED** [review]
This is a draft. 

What has been done so far: 
* Refactored the collator design which is now composable. Meaning you can call multiple collate_fns sequentially. This is helpful for instance for instruction tuning, where you first shift the targets for the autoregressive objective and then mask certiain tokens to disregard in the loss such as input from the user role. In this case, we would have two collate_fns being called one after the other. 

* We can specify now, which component to load in the app_state from a DCP checkpoint. This is important for continued pretraining or finetuning, where we don't want to use the previous optimizer and lr scheduler states. 

* Added an iterative mem map dataset implemenation. Instead of packing the samples, we take the index (stored in the pbin) to iterate over each sample. 

Todos:
* Add tests for the new functionality 
* Fix failing tests. 
* Verify correctness of the code.
